java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,660 @@
+    SemanticModel model = SemanticModel.createFor(tree, classLoader);
+
+    return new ScannedFile(tree, model);
+  }
+
+  public static class TaintSummary {
+
+    private final TaintSource result;
+    private final Set<TaintSource> callees;
+
+    public TaintSummary(TaintSource result, Set<TaintSource> callees) {
+      this.result = result;
+      this.callees = callees;
+    }
+
+    public TaintSource result() {
+      return result;
+    }
+
+    public Set<TaintSource> callees() {
+      return Collections.unmodifiableSet(callees);
+    }
+
+  }
+
+  public interface TaintSource {
+
+    boolean canBeTainted();
+    Set<TaintSource> directlyDependsOn();
+    Set<TaintSource> recursivelyDependsOn();
+    TaintSource unionWith(TaintSource other);
+
+  }
+
+  /**
+   * Read-from method parameter
+   * Read-from field
+   * Read-from method-call return value
+   */
+  public static class DirectTaintSource implements TaintSource {
+
+    private final String signature;
+    private final List<TaintSource> arguments;
+
+    public DirectTaintSource(String signature, List<TaintSource> arguments) {
+      this.signature = signature;
+      this.arguments = arguments;
+    }
+
+    @Override
+    public boolean canBeTainted() {
+      return true;
+    }
+
+    @Override
+    public Set<TaintSource> directlyDependsOn() {
+      return Collections.singleton(this);
+    }
+
+    @Override
+    public Set<TaintSource> recursivelyDependsOn() {
+      Set<TaintSource> result = new HashSet<>();
+      result.add(this);
+      for (TaintSource ts: arguments) {
+        result.addAll(ts.recursivelyDependsOn());
+      }
+      return Collections.unmodifiableSet(result);
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return IndirectTaintSource.union(this, other);
+    }
+
+    @Override
+    public String toString() {
+      return signature + (arguments.isEmpty() ? "" : "(" + Joiner.on(", ").join(arguments) + ")");
+    }
+
+    @Override
+    public int hashCode() {
+      return signature.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      DirectTaintSource other = (DirectTaintSource) obj;
+
+      return signature.equals(other.signature) &&
+        arguments.equals(other.arguments);
+    }
+
+  }
+
+  /**
+   * Literals (taint-free)
+   * Non-string operands (taint-free)
+   * String concatenation
+   * Multi-paths return values
+   * Multi-paths method argument taint conditions
+   */
+  public static class IndirectTaintSource implements TaintSource {
+
+    private Set<TaintSource> sources;
+
+    private IndirectTaintSource() {
+      this.sources = Collections.emptySet();
+    }
+
+    private IndirectTaintSource(TaintSource ts1, TaintSource ts2) {
+      Set<TaintSource> sources = new HashSet<>();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "sources" which hides the field declared at line 173. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AHiddenFieldCheck)